README
Name
    Catalyst::Plugin::InflateMore - Inflates symbols in application config

Synopsis
       package YourApp;

       use Catalyst qw(InflateMore ConfigLoader ...);

       # In your applications config file
       <appldir>__APPLDIR__</appldir>
       <binsdir>__BINSDIR__</binsdir>
       <libsdir>__LIBSDIR__</libsdir>
       <ctrldir>__appldir(var/etc)__</ctrldir>
       <dbasedir>__appldir(var/db)__</dbasedir>
       <logfile>__appldir(var/logs/server.log)__</logfile>
       <logsdir>__appldir(var/logs)__</logsdir>
       <root>__appldir(var/root)__</root>
       <rprtdir>__appldir(var/root/reports)__</rprtdir>
       <rundir>__appldir(var/run)__</rundir>
       <skindir>__appldir(var/root/skins)__</skindir>
       <tempdir>__appldir(var/tmp)__</tempdir>
       <vardir>__appldir(var)__</vardir>

Version
    This documents version v0.7.$Rev: 1 $ of Catalyst::Plugin::InflateMore

Description
    If symbols like *__MYSYMBOL__*, *__BINSDIR__*, or *__binsdir()__* are
    present in the application config they will be inflated to the
    appropriate directory paths if the corresponding lower case method name
    is defined in the inflation class

Configuration and Environment
    The *Plugin::InflateMore* attribute in the application config hash
    contains the name of the class whose methods will do the actual
    inflating

    Symbols should always use the forward slash as a path separator
    regardless of OS type, i.e. *__appldir(var/logs)__*

Subroutines/Methods
  setup
    Create an instance of the class that will do the inflating

  finalize_config
    Override Catalyst::Plugin::ConfigLoader method. Will inflate any symbols
    matching the patterns *__SYMBOL__* and *__symbol( value )__*

  _inflate_symbols
    Call the appropriate method to get the base path and append any
    arguments. Returns a Path::Class object representing the arguments
    passed

Diagnostics
    None

Dependencies
    Catalyst::Utils
    Data::Visitor::Callback
    Path::Class

Incompatibilities
    There are no known incompatibilities in this module

Bugs and Limitations
    There are no known bugs in this module. Please report problems to
    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Catalyst-Plugin-InflateMore.
    Patches are welcome

Acknowledgements
    Larry Wall - For the Perl programming language

Author
    Peter Flanigan, "<pjfl@cpan.org>"

License and Copyright
    Copyright (c) 2013 Peter Flanigan. All rights reserved

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic

    This program is distributed in the hope that it will be useful, but
    WITHOUT WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE



Changes
Revision history for Catalyst::Plugin::InflateMore

0.7.1     2013-08-21 00:05:15
          - Fixed metadata

0.6.4     2013-08-15 14:32:07
          - Updated toolchain
          - Added package signing
          - Added MetaProvides::Package to dist.ini
          - Converted from SVN to Git

0.5.2     2013-08-11 14:00:21
          - Added code from SVN repository
          - Added project files to master branch



inc/Bob.pm
# @(#)$Id: Bob.pm 83 2012-10-18 16:18:47Z pjf $

package Bob;

use strict;
use warnings;
use inc::CPANTesting;

sub whimper { print {*STDOUT} $_[ 0 ]."\n"; exit 0 }

BEGIN { my $reason; $reason = CPANTesting::should_abort and whimper $reason; }

use version; our $VERSION = qv( '1.5' );

use File::Spec::Functions;
use Module::Build;

sub new {
   my ($class, $p) = @_; $p ||= {}; $p->{requires} ||= {};

   my $perl_ver    = $p->{requires}->{perl} || 5.008_008;

   $] < $perl_ver and whimper "Perl minimum ${perl_ver}";

   my $module      = $p->{module} or whimper 'No module name';
   my $distname    = $module; $distname =~ s{ :: }{-}gmx;
   my $class_path  = catfile( q(lib), split m{ :: }mx, $module.q(.pm) );

   return __get_build_class( $p )->new
      ( add_to_cleanup     => __get_cleanup_list( $p, $distname ),
        build_requires     => $p->{build_requires},
        configure_requires => $p->{configure_requires},
        create_license     => 1,
        create_packlist    => 0,
        create_readme      => 1,
        dist_version_from  => $class_path,
        license            => $p->{license} || q(perl),
        meta_merge         => __get_resources( $p, $distname ),
        module_name        => $module,
        no_index           => __get_no_index( $p ),
        notes              => __get_notes( $p ),
        recommends         => $p->{recommends},
        requires           => $p->{requires},
        sign               => defined $p->{sign} ? $p->{sign} : 1, );
}

# Private functions

sub __get_build_class { # Which subclass of M::B should we create?
   my $p = shift; exists $p->{build_class} and return $p->{build_class};

   return Module::Build->subclass( code => q{
      use Pod::Select;

      sub ACTION_distmeta {
         my $self = shift;

         $self->notes->{create_readme_pod} and podselect( {
            -output => q(README.pod) }, $self->dist_version_from );

         return $self->SUPER::ACTION_distmeta;
      }
   } );
}

sub __get_cleanup_list {
   my $p = shift; my $distname = shift;

   return [ q(Debian_CPANTS.txt), "${distname}-*",
            map { ( q(*/) x $_ ).q(*~) } 0..5 ];
}

sub __get_git_repository {
   my ($info, $repo, $vcs); require Git::Class;

   $vcs = Git::Class::Worktree->new( path => q(.) )
      and $info = $vcs->git( q(remote) )
      and $repo = ($info !~ m{ \A file: }mx) ? $info : undef
      and return $repo;

   return;
}

sub __get_no_index {
   my $p = shift;

   return { directory => $p->{no_index_dir} || [ qw(examples inc t) ] };
}

sub __get_notes {
   my $p = shift; my $notes = exists $p->{notes} ? $p->{notes} : {};

   $notes->{create_readme_pod} = $p->{create_readme_pod} || 0;
   $notes->{stop_tests} = CPANTesting::test_exceptions( $p );

   return $notes;
}

sub __get_repository { # Accessor for the VCS repository information
   my $repo;

   -d q(.svn) and $repo = __get_svn_repository() and return $repo;
   -d q(.git) and $repo = __get_git_repository() and return $repo;

   return;
}

sub __get_resources {
   my $p         = shift;
   my $distname  = shift;
   my $tracker   = defined $p->{bugtracker}
                 ? $p->{bugtracker}
                 : q(http://rt.cpan.org/NoAuth/Bugs.html?Dist=);
   my $resources = $p->{resources} || {};
   my $repo;

   $tracker and $resources->{bugtracker} = $tracker.$distname;
   $p->{home_page} and $resources->{homepage} = $p->{home_page};
   $resources->{license} ||= q(http://dev.perl.org/licenses/);

   # Only get repository info when authoring a distribution
   -f q(MANIFEST.SKIP) and $repo = __get_repository
      and $resources->{repository} = $repo;

   return { resources => $resources };
}

sub __get_svn_repository {
   my ($info, $repo, $vcs); require SVN::Class;

   $vcs = SVN::Class::svn_dir( q(.) )
      and $info = $vcs->info
      and $repo = ($info->root !~ m{ \A file: }mx) ? $info->root : undef
      and return $repo;

   return;
}

1;

# Local Variables:
# mode: perl
# tab-width: 3
# End:


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.31 hacked
use strict;
use warnings;
sub _error { my $text = shift; print {*STDOUT} $text."\n"; exit 0 }
eval { require 5.008; }; _error( q(Perl minimum 5.008) ) if ($@);
use File::Spec::Functions;
my $found = 0;
for (@INC) { if (-f catfile( $_, q(version.pm) )) { $found = 1; last } }
_error( q(Version not installed) ) unless ($found);
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'Catalyst::Plugin::InflateMore',
          'VERSION_FROM' => 'lib/Catalyst/Plugin/InflateMore.pm',
          'PREREQ_PM' => {
                           'Catalyst' => '5.80007',
                           'Data::Visitor' => '0.08',
                           'Path::Class' => '0.16',
                           'Test::More' => '0.74',
                           'namespace::autoclean' => '0.05',
                           'parent' => '0.221',
                           'version' => '0.74'
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => []
        )
;


lib/Catalyst/Plugin/InflateMore.pm
# @(#)Ident: InflateMore.pm 2013-08-12 15:40 pjf ;

package Catalyst::Plugin::InflateMore;

use strict;
use warnings;
use namespace::autoclean;
use version; our $VERSION = qv( sprintf '0.7.%d', q$Rev: 1 $ =~ /\d+/gmx );

use Catalyst::Utils;
use Data::Visitor::Callback;
use Path::Class ();

my $KEY = 'Plugin::InflateMore'; my $SEP = '/';

sub setup {
   my ($self, @rest) = @_; $self->mk_classdata( '_inflator' );

   if (my $class = $self->config->{ $KEY }) {
      Catalyst::Utils::ensure_class_loaded( $class );
      $self->_inflator( $class->new( $self ) );
   }

   return $self->next::method( @rest );
}

sub finalize_config {
   my $self = shift;
   my $v    = Data::Visitor::Callback->new
      (  plain_value => sub {
            return unless defined $_;
            s{ __(.+?)\((.+?)\)__ }{$self->_inflate_symbols( $1, $2  )}egmx;
            s{ __(.+?)__          }{$self->_inflate_symbols( $1, q() )}egmx;
         } );

   $v->visit( $self->config );
   return;
}

# Private methods
sub _inflate_symbols {
   my ($self, $attr, @rest) = @_; $attr = lc $attr;

   $attr eq 'home' and return $self->path_to( $rest[ 0 ] );

   $self->_inflator->can( 'inflate' )
      and return $self->_inflator->inflate( $attr, @rest );

   my @parts = ($self->_inflator->$attr(), split m{ $SEP }mx, $rest[ 0 ]);
   my $path  = Path::Class::Dir->new( @parts );

   -d $path or $path = Path::Class::File->new( @parts );

   return $path->cleanup;
}

1;

__END__

=pod

=encoding utf8

=head1 Name

Catalyst::Plugin::InflateMore - Inflates symbols in application config

=head1 Synopsis

   package YourApp;

   use Catalyst qw(InflateMore ConfigLoader ...);

   # In your applications config file
   <appldir>__APPLDIR__</appldir>
   <binsdir>__BINSDIR__</binsdir>
   <libsdir>__LIBSDIR__</libsdir>
   <ctrldir>__appldir(var/etc)__</ctrldir>
   <dbasedir>__appldir(var/db)__</dbasedir>
   <logfile>__appldir(var/logs/server.log)__</logfile>
   <logsdir>__appldir(var/logs)__</logsdir>
   <root>__appldir(var/root)__</root>
   <rprtdir>__appldir(var/root/reports)__</rprtdir>
   <rundir>__appldir(var/run)__</rundir>
   <skindir>__appldir(var/root/skins)__</skindir>
   <tempdir>__appldir(var/tmp)__</tempdir>
   <vardir>__appldir(var)__</vardir>

=head1 Version

This documents version v0.7.$Rev: 1 $ of L<Catalyst::Plugin::InflateMore>

=head1 Description

If symbols like I<__MYSYMBOL__>, I<__BINSDIR__>, or I<__binsdir()__>
are present in the application config they will be inflated to the
appropriate directory paths if the corresponding lower case method name
is defined in the inflation class

=head1 Configuration and Environment

The I<Plugin::InflateMore> attribute in the application config hash
contains the name of the class whose methods will do the actual
inflating

Symbols should always use the forward slash as a path separator regardless
of OS type, i.e. I<__appldir(var/logs)__>

=head1 Subroutines/Methods

=head2 setup

Create an instance of the class that will do the inflating

=head2 finalize_config

Override L<Catalyst::Plugin::ConfigLoader> method. Will inflate any
symbols matching the patterns I<__SYMBOL__> and I<__symbol( value )__>

=head2 _inflate_symbols

Call the appropriate method to get the base path and append any
arguments. Returns a L<Path::Class> object representing the arguments
passed

=head1 Diagnostics

None

=head1 Dependencies

=over 3

=item L<Catalyst::Utils>

=item L<Data::Visitor::Callback>

=item L<Path::Class>

=back

=head1 Incompatibilities

There are no known incompatibilities in this module

=head1 Bugs and Limitations

There are no known bugs in this module. Please report problems to
http://rt.cpan.org/NoAuth/Bugs.html?Dist=Catalyst-Plugin-InflateMore.
Patches are welcome

=head1 Acknowledgements

Larry Wall - For the Perl programming language

=head1 Author

Peter Flanigan, C<< <pjfl@cpan.org> >>

=head1 License and Copyright

Copyright (c) 2013 Peter Flanigan. All rights reserved

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>

This program is distributed in the hope that it will be useful,
but WITHOUT WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE

=cut

# Local Variables:
# mode: perl
# tab-width: 3
# End:
# vim: expandtab shiftwidth=3:


Build.PL
# @(#)$Ident: DZ_Build.PL 2013-08-15 13:04 pjf ;

use strict;
use warnings;

use Module::Build;

sub whimper { print {*STDOUT} $_[ 0 ]."\n"; exit 0 }

my $perl_ver = 5.008;
my $name     = 'Catalyst-Plugin-InflateMore'; $name =~ s{ - }{::}gmx;

$] >= $perl_ver or $ENV{PERL5_CPANPLUS_IS_VERSION}
   or whimper "Minimum required Perl version is ${perl_ver}";

Module::Build->subclass( code => do { local $/; <DATA> } )->new(
  'module_name' => $name,
  'notes'       => { min_perl_version => $perl_ver },
  'build_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'configure_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'requires' => {
    'Catalyst' => '5.90010',
    'Data::Visitor' => '0.08',
    'Path::Class' => '0.16',
    'namespace::autoclean' => '0.05',
    'perl' => '5.008',
    'version' => '0.88'
  },

)->create_build_script;

exit 0;

__DATA__

sub _normalize_prereqs { # Dynamic config
   my $self = shift; my $p = $self->SUPER::_normalize_prereqs; return $p;
}
